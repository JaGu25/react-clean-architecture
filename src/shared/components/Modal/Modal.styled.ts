
import styled from 'styled-components'
import { colors, fontSizes } from '../../utils/styles.utils'

const ModalContainer = styled.div`
    position: absolute;
    top: 0;
    left: 0;
    background-color: rgba(0, 0, 0, 0.3);
    width: 100%;
    min-height: 100vh;
    display: grid;
    align-items: flex-end;
    overflow: hidden;
    transition: all 0.3s;
    z-index: 1000;
    visibility: hidden;
    opacity: 0;
    
    &.show {
        visibility: visible;
        opacity: 1;

        > div {
            transform: scale(1) translateY(0);
        }
    }
`

const ModalContent = styled.div`
    position: relative;
    background-color: ${colors.white};
    text-align: center;
    padding: 3rem 2rem;
    border-radius: 1rem 1rem 0 0;
    transition: all 0.3s;
    margin: auto;
    min-width: 380px;
    border-radius: 1.25rem;
    transform: scale(.5) translateY(10%);
`

const ModalClose = styled.i`
    position: absolute;
    top: 10px;
    right: 15px;
    font-size: ${fontSizes.lg};
    color: ${colors.primary};
    cursor: pointer;
`

const ModalStyled = {
    ModalContainer,
    ModalContent,
    ModalClose
}

export default ModalStyled
