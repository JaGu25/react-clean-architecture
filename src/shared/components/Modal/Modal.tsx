import React, { Children, Dispatch, SetStateAction, useRef } from 'react'
import useOutsideElement from '../../hooks/useClickOutsideElement'
import useKeypress from '../../hooks/useKeypress'
import ModalStyled from './Modal.styled'

interface IProps {
    showModal: boolean
    handleModal: Dispatch<SetStateAction<boolean>>
    children: React.ReactElement
}

const Modal: React.FC<IProps> = ({
    showModal,
    handleModal,
    children
}) => {

    const wrapperRef = useRef<HTMLDivElement>(null)

    useKeypress("Escape", () => handleModal(false))
    useOutsideElement(wrapperRef, () => handleModal(false))

    return (
        <ModalStyled.ModalContainer className={`${showModal && ('show')}`}>
            <ModalStyled.ModalContent ref={wrapperRef}>
                <ModalStyled.ModalClose
                    data-testid="closemodal"
                    className='fa-solid fa-xmark'
                    onClick={() => handleModal(false)} />
                {children}
            </ModalStyled.ModalContent>
        </ModalStyled.ModalContainer>
    )
}

export default Modal
