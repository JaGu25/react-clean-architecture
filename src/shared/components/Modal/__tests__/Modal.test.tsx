import React, { useState as useStateMock } from "react"
import { fireEvent, render, screen } from "@testing-library/react"
import Modal from "../Modal"

const click = jest.fn()

jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useState: jest.fn(),
}));

describe('Button Component', () => {
    it('should render without crash', () => {
        const { container } = render(
            <Modal
                handleModal={click}
                showModal={true}>
                <h1>Modal contenido</h1>
            </Modal>
        )
        expect(container).toMatchSnapshot()
    })

    it('should show modal', async () => {
        render(
            <Modal
                handleModal={click}
                showModal={true}>
                <h1>Modal contenido</h1>
            </Modal>
        )

        expect(await screen.findByText('Modal contenido')).toBeInTheDocument
    })

    it('should close modal', async () => {

        const handleModal = jest.fn()

        render(
            <Modal
                handleModal={handleModal}
                showModal={true}>
                <h1>Modal contenido</h1>
            </Modal>
        )

        const closeModalButton = await screen.findByTestId('closemodal')
        fireEvent.click(closeModalButton)
        expect(handleModal).toBeCalled()
    })
})