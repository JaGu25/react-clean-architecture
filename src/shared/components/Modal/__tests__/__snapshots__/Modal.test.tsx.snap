// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component should render without crash 1`] = `
<div>
  <div
    class="sc-bczRLJ NFeCt show"
  >
    <div
      class="sc-gsnTZi gVgvKC"
    >
      <i
        class="sc-dkzDqf dmOPiU fa-solid fa-xmark"
        data-testid="closemodal"
      />
      <h1>
        Modal contenido
      </h1>
    </div>
  </div>
</div>
`;
